src/vm/database/marf.rs

```diff
@@ -0,0 +1,139 @@
+///   for use as a K/V store for ClarityDB or the AnalysisDB.
+/// The Clarity VM and type checker do not "know" to begin/commit the block they are currently processing:
+///   each instantiation of the VM simply executes one transaction. So the block handling
+///   loop will need to invoke these two methods (begin + commit) outside of the context of the VM.
+///   NOTE: Clarity will panic if you try to execute it from a non-initialized MarfedKV context.
+///   (See: vm::tests::with_marfed_environment()) 
+pub struct MarfedKV {
+    chain_tip: BlockHeaderHash,
+    marf: MARF,
+    // Since the MARF only stores 32 bytes of value,
+    //   we need another storage
+    side_store: Box<KeyValueStorage>
+}
+
+#[cfg(test)]
+pub fn temporary_marf() -> MarfedKV {
+    use std::env;
+    use rand::Rng;
+    use util::hash::to_hex;
+    use std::collections::HashMap;
+
+    let mut path = env::temp_dir();
+    let random_bytes = rand::thread_rng().gen::<[u8; 32]>();
+    path.push(to_hex(&random_bytes));
+
+    let mut marf = MARF::from_path(path.to_str().expect("Inexplicably non-UTF-8 character in filename"))
+        .unwrap();
+    let side_store = Box::new(HashMap::new());
+
+    let chain_tip = marf.chain_tips().get(0).cloned()
+        .unwrap_or_else(|| TrieFileStorage::block_sentinel());
+
+    MarfedKV { chain_tip, marf, side_store }
+}
+
+/// If chain_tip is None, this will try to figure out a reasonable value for a starting
+///   chain_tip. If the MARF already has some chain_tips, then it selects ordinal 0.
+///   Otherwise, it uses the block_sentinel.
+pub fn sqlite_marf(path_str: &str, chain_tip: Option<BlockHeaderHash>) -> Result<MarfedKV> {
+    let mut path = PathBuf::from(path_str);
+    std::fs::create_dir_all(&path)
+        .map_err(|err| InterpreterError::FailedToCreateDataDirectory)?;
+
+    path.push("marf");
+    let marf_path = path.to_str()
+        .ok_or_else(|| InterpreterError::BadFileName)?
+        .to_string();
+
+    path.pop();
+    path.push("data.sqlite");
+    let data_path = path.to_str()
+        .ok_or_else(|| InterpreterError::BadFileName)?
+        .to_string();
+
+    let side_store = Box::new(SqliteConnection::initialize(&data_path)?);
+    let mut marf = MARF::from_path(&marf_path)
+        .map_err(|err| InterpreterError::MarfFailure(IncomparableError{ err }))?;
+
+    let chain_tip = chain_tip
+        .or_else(|| marf.chain_tips().get(0).cloned())
+        .unwrap_or_else(|| TrieFileStorage::block_sentinel());
+
+
+    Ok( MarfedKV { chain_tip, marf, side_store } )
+}
+
+impl MarfedKV {
+    pub fn begin(&mut self, current: &BlockHeaderHash, next: &BlockHeaderHash) {
+        self.marf.begin(current, next)
+            .unwrap();
+        self.chain_tip = next.clone();
+    }
+    pub fn commit(&mut self) {
+        self.marf.commit()
+            .unwrap()
+    }
+    pub fn chain_tips(&mut self) -> Vec<BlockHeaderHash> {
+        self.marf.chain_tips()
+    }
```

Comment:
There should probably be something like `get_longest_chain_tip()` or `get_chain_length(&self, tip: &BlockHeaderHash) -> u64`.  Applies to the MARF as well.